### DISCUSSION TITLE
Raspberry Pico and MHS 4001

### DISCUSSION THREAD

**Discussion by dietmarczekay** [Bodmer/TFT_eSPI#2255] [General]
Hi,
I'm completely lost. I try to get the MHS4001 to work with my pico. 
[http://www.lcdwiki.com/File:MHS4001-02.jpg]
The Setup file is

```
#define USER_SETUP_ID 162
#define ST7796_DRIVER

#define TFT_MISO  0
#define TFT_MOSI  3
#define TFT_SCLK  2
#define TFT_CS   20  // Chip select control pin
#define TFT_DC   18  // Data Command control pin
#define TFT_RST  19  // Reset pin (could connect to Arduino RESET pin)
//#define TFT_BL     // LED back-light

//#define TOUCH_CS 21     // Chip select pin (T_CS) of touch screen


#define LOAD_GLCD   // Font 1. Original Adafruit 8 pixel font needs ~1820 bytes in FLASH
#define LOAD_FONT2  // Font 2. Small 16 pixel high font, needs ~3534 bytes in FLASH, 96 characters
#define LOAD_FONT4  // Font 4. Medium 26 pixel high font, needs ~5848 bytes in FLASH, 96 characters
#define LOAD_FONT6  // Font 6. Large 48 pixel font, needs ~2666 bytes in FLASH, only characters 1234567890:-.apm
#define LOAD_FONT7  // Font 7. 7 segment 48 pixel font, needs ~2438 bytes in FLASH, only characters 1234567890:-.
#define LOAD_FONT8  // Font 8. Large 75 pixel font needs ~3256 bytes in FLASH, only characters 1234567890:-.
//#define LOAD_FONT8N // Font 8. Alternative to Font 8 above, slightly narrower, so 3 digits fit a 160 pixel TFT
#define LOAD_GFXFF  // FreeFonts. Include access to the 48 Adafruit_GFX free fonts FF1 to FF48 and custom fonts

#define SMOOTH_FONT

#define SPI_FREQUENCY  20000000
#define SPI_READ_FREQUENCY  20000000
#define SPI_TOUCH_FREQUENCY  2500000
```
Upload works, I get messages on the serial console but the TFT stays white. No Text, no black background.
What do I miss?

**Comment by dietmarczekay** [comment]
In the readme I red the MHS-dislpay is a good starting point... Seems so, not with the pico.

The ino-file looks like
```
/*  
 Test the tft.print() viz the libraries embedded write() function
 This sketch used font 2, 4, 7
 
 Make sure all the required fonts are loaded by editing the
 User_Setup.h file in the TFT_eSPI library folder.
  #########################################################################
  ###### DON'T FORGET TO UPDATE THE User_Setup.h FILE IN THE LIBRARY ######
  #########################################################################
 */

#include <SPI.h>
#include <TFT_eSPI.h> // Hardware-specific library

TFT_eSPI tft = TFT_eSPI();       // Invoke custom library

void setup(void) {
  
  // initialize digital pin LED_BUILTIN as an output.
  pinMode(15, OUTPUT);
  delay(1000);                      // wait for a second
  Serial.begin(9600);
  Serial.println("Hello, start!");

  tft.init();
  tft.setRotation(2);
  tft.fillScreen(TFT_BLACK);
  tft.setTextSize(1);
  tft.setTextColor(TFT_WHITE);
  tft.setCursor(0, 0);
}

// the loop function runs over and over again forever
void loop() {

  digitalWrite(15, HIGH);  // turn the LED on (HIGH is the voltage level)
  delay(1000);                      // wait for a second
  digitalWrite(15, LOW);   // turn the LED off by making the voltage LOW
  delay(1000);                      // wait for a second
  tft.fillScreen(TFT_BLACK);
  tft.setTextSize(2);
  tft.setTextColor(TFT_WHITE);
  tft.setCursor(40, 5);
  tft.println(F("Pico"));
}
```

**Comment by Bodmer** [comment]
The display is designed for the Raspberry Pi and thus has a custom interface design. To select the correct interface type you must add this line to the setup file:

#define RPI_DISPLAY_TYPE

**Comment by dietmarczekay** [comment]
Great, that solved the problem.
Where can i find the information to distinguish between display types?

**Reply by Bodmer** [reply]
See the [default setup](https://github.com/Bodmer/TFT_eSPI/blob/master/User_Setup.h) file for information on settings.

**Comment by Bodmer** [comment]
That particular display is very unusual in that it can be clocked at 125MHz. The best performance (on the RP2040) without overclocking will be with a CPU freuency set to 125MHz and the SPI clock rate set to 62.5MHz:

#define SPI_FREQUENCY  62500000

With Earle Philhower's board package you can change the CPU speed (aka frequency) in the Arduino IDE "Tools" menu. At the default core package CPU speed of 133MHz and above the maximum SPI clock rate drops to 24MHz for the reasons I detailed here:

https://github.com/earlephilhower/arduino-pico/discussions/415

Note that if you over-clock the processor to 250MHz and include the clock switch code (in above link) at the start of setup(), you can then specifiy a maximum SPI frequency of half the CPU speed, in this case 125MHz:

#define SPI_FREQUENCY  125000000

This then gives a very good display update speed (circa 45fps for clear screen which is the highest of any 320x480 SPI interface display I have tested), so this is why it is a recommended display.  At some point I will add PIO support for that particular display type to improve the rendering performance further.
